src/parser/jobs/whm/modules/Lilies.tsx

```diff
@@ -0,0 +1,159 @@
+import {Plural, Trans} from '@lingui/react'
+import Color from 'color'
+import {DataLink} from 'components/ui/DbLink'
+import {ActionKey} from 'data/ACTIONS'
+import {Event} from 'event'
+import {filter} from 'parser/core/filter'
+import {dependency} from 'parser/core/Injectable'
+import {CounterGauge, Gauge as CoreGauge, TimerGauge} from 'parser/core/modules/Gauge'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+import React from 'react'
+
+const LILY_MAX_STACKS = 3
+const LILY_INTERVAL_600 = 30000
+const LILY_INTERVAL_610 = 20000
+
+const MISERY_COST = 3
+
+const GRAPH_DISPLAY_SETTINGS = {
+	LILY_GAUGE_FADE: 0.25,
+	LILY_TIMER_FADE: 0.75,
+	BLOODLILY_GAUGE_FADE: 0.5,
+	LILY_GAUGE_COLOR: Color('#4f73b5'),
+	BLOODLILY_GAUGE_COLOR: Color('#b52d6c'),
+}
+
+export class Lilies extends CoreGauge {
+
+	@dependency private suggestions!: Suggestions
+
+	private lilyConsumers: ActionKey[] = [
+		'AFFLATUS_RAPTURE',
+		'AFFLATUS_SOLACE',
+	]
+
+	private bloodLilyConsumers: ActionKey[] = [
+		'AFFLATUS_MISERY',
+	]
+
+	private lilyInterval = this.parser.patch.before('6.1') ? LILY_INTERVAL_600 : LILY_INTERVAL_610
+
+	private lilyGauge = this.add(new CounterGauge({
+		maximum: LILY_MAX_STACKS,
+		initialValue: 0,
+		graph: {
+			label: <Trans id="whm.gauge.lily.stacks.label">Lily</Trans>,
+			color: GRAPH_DISPLAY_SETTINGS.LILY_GAUGE_COLOR.fade(GRAPH_DISPLAY_SETTINGS.LILY_GAUGE_FADE),
+		},
+	}))
+
+	private lilyTimer = this.add(new TimerGauge({
+		maximum: this.lilyInterval,
+		onExpiration: this.onLilyGeneration.bind(this),
+		graph: {
+			label: <Trans id="whm.gauge.lily.timer.label">Lily Timer</Trans>,
+			color: GRAPH_DISPLAY_SETTINGS.LILY_GAUGE_COLOR.fade(GRAPH_DISPLAY_SETTINGS.LILY_TIMER_FADE),
+		},
+
+	}))
+
+	private bloodLilyGauge = this.add(new CounterGauge({
+		maximum: LILY_MAX_STACKS,
+		initialValue: 0,
+		graph: {
+			label: <Trans id="whm.gauge.bloodlily.stacks.label">Blood Lily</Trans>,
+			color: GRAPH_DISPLAY_SETTINGS.BLOODLILY_GAUGE_COLOR.fade(GRAPH_DISPLAY_SETTINGS.BLOODLILY_GAUGE_FADE),
+		},
+	}))
+
+	override initialise() {
+		super.initialise()
+
+		const playerFilter = filter<Event>().source(this.parser.actor.id)
+		this.addEventHook(playerFilter.type('action').action(this.data.matchActionId(this.lilyConsumers)), this.onLilySpend)
+		this.addEventHook(playerFilter.type('action').action(this.data.matchActionId(this.bloodLilyConsumers)), () => this.bloodLilyGauge.spend(MISERY_COST))
+		this.addEventHook('complete', this.onComplete)
+		this.lilyTimer.start()
+	}
+
+	private onLilyGeneration() {
+		this.lilyGauge.generate(1)
+		if (!this.lilyGauge.capped) {
+			this.lilyTimer.start()
+		}
+	}
+
+	private onLilySpend() {
+		this.bloodLilyGauge.generate(1)
+		if (this.lilyTimer.expired) {
+			this.lilyTimer.start()
+		}
+		this.lilyGauge.spend(1)
```

Comment:
Nit; easier to read this way. Looks like indentation was borked (spaces rather than tabs) too.
```suggestion
		this.bloodLilyGauge.generate(1)
		this.lilyGauge.spend(1)
		if (this.lilyTimer.expired) {
			this.lilyTimer.start()
		}
```